streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -253,33 +288,103 @@ public synchronized State state() {
         return Collections.unmodifiableMap(metrics.metrics());
     }
 
-    private final class StreamStateListener implements StreamThread.StateListener {
-
+    /**
+     * Class that handles stream thread transitions
+     */
+    final class StreamStateListener implements StreamThread.StateListener {
         private final Map<Long, StreamThread.State> threadState;
+        private GlobalStreamThread.State globalThreadState;
 
-        StreamStateListener(Map<Long, StreamThread.State> threadState) {
+        StreamStateListener(final Map<Long, StreamThread.State> threadState,
+                            final GlobalStreamThread.State globalThreadState) {
             this.threadState = threadState;
+            this.globalThreadState = globalThreadState;
+        }
+
+        /**
+         * If all threads are dead set to ERROR
+         */
+        private void checkAllThreadsDeadAndSetError() {
```

Comment:
I think the lock grabbing hierarchy of instance `stateLock` and thread `stateLock` is still vulnerable to deadlocks. For example:

thread 1: `StreamThread.setState` --> grab `stateLock` of thread --> `stateListener.onChange` --> `checkAllThreadsDeadAndSetError` --> trying to grab `stateLock` of instance

thread 2: `KafkaStreams.setState` --> grab `stateLock` of instance -> user-specified `stateListener.onChange` --> user calls `KafkaStreams.close` inside that callback --> wait for all the stream-thread to shutdown, but thread 1 is blocked on grabbing the instance `stateLock`.